72_config_and_simplify_tls.md

```diff
@@ -0,0 +1,358 @@
+<!--
+<!--
+**Note:** When your enhancement is complete, all of these comment blocks should be removed.
+
+To get started with this template:
+
+- [ ] **Create an issue in keylime/enhancements**
+  When filing an enhancement tracking issue, please ensure to complete all
+  fields in that template.  One of the fields asks for a link to the enhancement.  You
+  can leave that blank until this enhancement is made a pull request, and then
+  go back to the enhancement and add the link.
+- [ ] **Make a copy of this template.**
+ name it `NNNN-short-descriptive-title`, where `NNNN` is the issue number (with no
+  leading-zero padding) assigned to your enhancement above.
+- [ ] **Fill out this file as best you can.**
+  At minimum, you should fill in the "Summary", and "Motivation" sections.
+  These should be easy if you've preflighted the idea of the enhancement with the
+  appropriate SIG(s).
+- [ ] **Merge early and iterate.**
+  Avoid getting hung up on specific details and instead aim to get the goals of
+  the enhancement clarified and merged quickly.  The best way to do this is to just
+  start with the high-level sections and fill out details incrementally in
+  subsequent PRs.
+-->
+# enhancement-72: Move configuration to TOML and Simplify TLS Setup
```

Comment:
Why TOML? We are using INI, and I do not see any limitation. There are parsers for both Rust (including the for the serde crate) and Python.

Do we need any extra feature from TOML, like arrays?

Reply:
Yes, there is a current open issue about this: https://github.com/keylime/keylime/issues/895

Reply:
Lets pick one for array then (`key=value1,value2` for example) and use `converters` (https://docs.python.org/3/library/configparser.html#configparser-objects)

IMHO there is more value in normalizing the name of the elements in the config file, than in moving into TOML.

For example, using TOML `include = []` for composition is sub-optimal. The inclusion is explicit and is set in the root config file.

That is not what it is usual when configuring service, that is more in the tradition of creating `service_name.d/` directories, where the overwrites lives (apache, sshd, systemd, and many others work like that). This way of splitting the config file have also more value than using TOML for type normalization.
 
I understand that TOML is of value in Rust, because if you are using serde _and_ you are deserialzing into a `struct`, you will have type validation for free. I would be fine to use TOML for Rust, but there is no gain in using it in Python; there is no data validation: you will read in a dictionary.

Reply:
> I'm against using different formats for Python and rust, I prefer to keep current format (with normalized names) than having different formats depending on the implementation.

I agree, same configuration format for everything component.

Reply:
> That is not what it is usual when configuring service, that is more in the tradition of creating `service_name.d/` directories, where the overwrites lives (apache, sshd, systemd, and many others work like that). This way of splitting the config file have also more value than using TOML for type normalization.

I agree with this as well. The more similar we can be to other systems the lower barrier to entry and the easier it will be to productize, manage with config management tools, etc.

Is it possible to use TOML in a `conf.d` style setup?

Reply:
> For example, using TOML include = [] for composition is sub-optimal. The inclusion is explicit and is set in the root config file.

I'll switch to a `keylime.d` approach in the next iteration.


I agree that there is no immediate gain for the Python code switching to TOML, but we don't want to have two different formats for different parts of Keylime. The issue with the current INI code is that there is no clear way on how to encode booleans, ints, lists etc. With TOML we have this clearly defined in the spec.

Reply:
> I agree that there is no immediate gain for the Python code switching to TOML, but we don't want to have two different formats for different parts of Keylime.

I totally understand this, but I think that is a good tradeoff to have different formats (more later)

> The issue with the current INI code is that there is no clear way on how to encode booleans, ints, lists etc. With TOML we have this clearly defined in the spec.

This is true, but not quite. It is true that there is in TOML a fixed way of representing a boolean and a list, but there is no schema validation. So the gain is really minimal in Python: you still needs to provide a validator to make sure that, for example, the parameter `registrar_port` is an integer, and `tls_dir` is an string.

But this can be better in Rust, where you can define a `struct` that is used by serde to deserialize the TOML document. If the data in the config file does not match with the one in the `struct` it will complain. Note that serde can deserialize into a dict, so if we use this option we still need to provide a validator in the code.

So again, there is a tradeoff here:

Using INI has some costs:
  * we need to validate the input in Python, and pbly in Rust (except if we use serde with an struct)
  * formalize how the boolean and the list will be expressed (at least for lists, as I think that both Python and Rust parsers for INI understand the usual boolean reps for INI)

But we have for free:
 * the split of different files per sections
 * the merge of /usr/etc with /etc - already implemented
 * the split in snippets that live in keylime.d

Using TOML have other costs:
 * re-implement the config reader to use a different libraries in Python and Rust
 * we still need to validate the input in Python, and pbly in Rust (except if we use serde with an struct)
 * we need to implement the dictionary merge of different files in Python and Rust (for /etc and /usr/etc, and for keylime.d)
 * we need to provide a migration tool

But we have for free:
 * a way to indicate how is a boolean and how is a list

I hope not to be too biased here.

Maybe we can have a win if we focus on the key / value normalization in the RFC that is where there is the most gain.

I am still thinking that we should pivot the efforts in providing all the new and cool features in the Rust agent, and stop developing the Python one. Maybe if we do this, we will see clearly that the correct decision latter will be to use TOML only for the Rust agent, because the Python one will not be improved anymore.

But I am totally fine if we decide another thing.

Reply:
Thanks @aplanas for detailed breakdown of the tradeoffs. 

I actually agree with you that if we put the Python agent officially maintenance mode there are mainly downsides moving to TOML in the rest of the Python code. I think that the rust and Python agent configs will starting to diverge in the future and normally the agent is not deployed with the server components, so using another format for the rust agent should be fine.

What I will change:

* Keep the INI format, but add normalization to booleans and lists and enforce them
* Split the configuration files and put them into `/etc/keylime/` (or for distributions in `/usr/etc/keylime`)
* Allow overrides using `/etc/keylime.d/`
* Add a note that the agent config for the rust agent is no longer tracking the Python configuration

Reply:
@THS-on Could you please update the text to reflect these decisions?

Let's move this proposal forward.

Reply:
@ansasaki Yes, I'll try to do that tomorrow.

Reply:
Done